File: TSP/Arduino/TSP/TSP.cpp
#include <Arduino.h> 
#include "tsp.h"

/*
 * 公共函数
 */
char getChar(uint8_t data, int isHigh){
	char c;
	if(isHigh!=0){
		// high 4 bit
		c = (data>>4) & 0x0F;
	}else{
		c = data & 0x0F;
	}
	if((uint8_t)c < 10){
		c += 48;
	}else{
		c += 55;
	}
	return c;
}
int8_t getFrameString(uint8_t * data, int data_len, char * char_buf[96]){
	int strlen = 0;
	for(int i=0;i<data_len;i++){
		(*char_buf)[i*3] = getChar(data[i], 1);
		(*char_buf)[i*3+1] = getChar(data[i], 0);
		(*char_buf)[i*3+2] = ' ';
		strlen +=3;
	}
	(*char_buf)[strlen-1] = '\0';
	return strlen;
}
/*
 * 类内部
 */
DataFrame::DataFrame()
{
	//constructor
	start[0] = 0x7E;
	start[1] = 0x33;
	
	msg_type = 0; // 1 for data, 2 for operation
	// payload
	sub_type = 0;
	int i=0;
	for(i=0;i<4;i++){
		data1[i] = 0;
		data2[i] = 0;
		data3[i] = 0;
	}
	
   
	end = 0x7e;
}

uint32_t DataFrame::double2int32(double data){
	uint32_t result =0;
	result = data*100/1;
	return result;
}

/*
 * 设定操作
 */
void DataFrame::setOperationDF(uint8_t _sub_type, uint32_t operation_1, 
                      uint32_t operation_2, uint32_t operation_3){
	msg_type = 2;
	sub_type = _sub_type;
	int i;
	for(i=0; i<4; i++){
		data1[i]=(operation_1>>(i*8)) & 0xff;
		data2[i]=(operation_2>>(i*8)) & 0xff;
		data3[i]=(operation_3>>(i*8)) & 0xff;
	}
}


void DataFrame::setDataDF(uint8_t _sub_type, uint32_t data_1, 
					uint32_t data_2, uint32_t data_3){
	msg_type = 1;
	sub_type = _sub_type;
	int i;
	for(i=0; i<4; i++){
		data1[i]=(data_1>>(i*8)) & 0xff;
		data2[i]=(data_2>>(i*8)) & 0xff;
		data3[i]=(data_3>>(i*8)) & 0xff;
	}
}

void DataFrame::setDataByFormatByteStream(uint8_t * byte_stream){
	msg_type = byte_stream[2];
	sub_type = byte_stream[3];
	for(int i=0; i<4; i++){
		data1[i]= byte_stream[4+i];
		data2[i]= byte_stream[8+i];
		data3[i]= byte_stream[12+i];
	}
}

void DataFrame::setDataByOriginByteStream(uint8_t * ori_byte_stream, int len){
	// 起始验证
	if(ori_byte_stream[0]!=0x7E|| ori_byte_stream[1] != 0x33){
			return;
	}
	// rid off 转义
	uint8_t result_stream[17]; //V1.0版协议的总长度
	int aim_pointer = 0;
	int ori_pointer = 0;
	for(ori_pointer = 0; ori_pointer<len;){
		if(ori_byte_stream[ori_pointer]==0x7D){  //遇到7D
			if(ori_byte_stream[ori_pointer+1]==0x5E || ori_byte_stream[ori_pointer+1]==0x5D){
				result_stream[aim_pointer] = ori_byte_stream[ori_pointer+1] ^ 0x20;
				aim_pointer++;
				ori_pointer+=2;
			}else{
				result_stream[aim_pointer] = ori_byte_stream[ori_pointer];
				aim_pointer++;
				ori_pointer++;
			}
		}else{
			result_stream[aim_pointer] = ori_byte_stream[ori_pointer];
			aim_pointer++;
			ori_pointer++;
		}
	}
	if(aim_pointer==17){
		// 长度正确，解析完成
		setDataByFormatByteStream(result_stream);
	}
}


/*
 * 获取数据
 */	
uint32_t DataFrame::getValueInt(uint8_t index){
	uint32_t value=0;
	uint8_t * data;
	switch(index){
		case 1:
			data = data1;
			break;
		case 2:
			data = data2;
			break;
		case 3:
			data = data3;
			break;
		default:
			return 0;
	}
	int i;

	for(i=0;i<4;i++){
		value = value << 8;
		value += data[3-i] & 0xff;
	}

	return value;
}

/*
 * 获取数据
 */

double DataFrame::getValueDouble(uint8_t index){
	uint32_t tempValue = getValueInt(index);
	double result = tempValue/100.0;
	return result;
}

int pushFormatData(uint8_t ** aim_array, int index, uint8_t bytedata){
	int len = 0;
	if(bytedata==0x7E){
		(*aim_array)[index] = 0x7D;
		(*aim_array)[index+1] = 0x5E;
		len = 2;
	}else if(bytedata==0x7D){
		(*aim_array)[index] = 0x7D;
		(*aim_array)[index+1] = 0x5D;
		len = 2;
	}else{
		(*aim_array)[index] = bytedata;
		len = 1;
	}
	return len;
}
uint8_t DataFrame::makeSendData(uint8_t * databuf[32]){ // TODO 32 max
	//uint8_t senddata[32];
	uint8_t len = 0;
	//start
	(*databuf)[0] = start[0];
	(*databuf)[1] = start[1];
	len = 2;
	// payload
	
	int end_flag = 0;
	for(int i=2,j=0; end_flag<1;){
		int addLen = 0;
		if(j<1){
			addLen = pushFormatData(databuf, i, msg_type);
		}else if(j<2){
			addLen = pushFormatData(databuf, i, sub_type);
		}else if(j<6){
			addLen = pushFormatData(databuf, i, data1[j-2]);
		}else if(j<10){
			addLen = pushFormatData(databuf, i, data2[j-6]);
		}else if(j<14){
			addLen = pushFormatData(databuf, i, data3[j-10]);
			if(j == 13){
				end_flag = 1;
			}
		}
		j ++;
		i += addLen;
		len += addLen;
	}
	
	//end 
	(*databuf)[len] = end;
	len++;
	
	//result) = senddata;
	
	return len;	
}

File: TSP/Python/TSP/TSP.py
# -*- coding: utf-8 -*-

class DataFrame:

    def __init__(self):
        data = [0x7e, 0x33]     #[0][1]
        self.start = bytearray(data)

        self.msg_type = 0            #[2]
        self.sub_type = 0            #[3]
        
        data = [0,0,0,0]
        self.data1 = bytearray(data) #[4][5][6][7]
        self.data2 = bytearray(data) #[8][9][10][11]
        self.data3 = bytearray(data) #[12][13][14][15]
        
        data = 0x7e        
        self.end = data              #[16]
    
    def setDataByFormatByteSream(self, _mid_byte_stream):
        self.msg_type = _mid_byte_stream[2]
        self.sub_type = _mid_byte_stream[3]
        self.data1 = _mid_byte_stream[4:8]
        self.data2 = _mid_byte_stream[8:12]
        self.data3 = _mid_byte_stream[12:16]
        self.end = _mid_byte_stream[16]
        
    def setDataByByteStream(self, rcv_data_buf):
        result = False
        if(len(rcv_data_buf)<17):
            return result
        elif rcv_data_buf[0]!=0x7e or rcv_data_buf[1]!=0x33:
            return result
        else:
            mid_byte_stream = bytearray()
            trans_flag = False
            for inbyte in rcv_data_buf:
                if inbyte == 0x7D:
                    trans_flag = True
                elif inbyte == 0x5E or inbyte == 0x5D:
                    if trans_flag == True:
                        mid_byte_stream.append(inbyte ^ 0x20)
                    else:
                        mid_byte_stream.append(inbyte)
                    trans_flag = False
                else:
                    mid_byte_stream.append(inbyte)
                    trans_flag = False
            # end of transfer
            if len(mid_byte_stream) == 17:
                print "get a right frame" #TS.Test
                self.setDataByFormatByteSream(mid_byte_stream)
                return True
    
    def getValueInt(self, _index):
        intValue = 0
        if _index ==1:
            data = self.data1
        elif _index==2:
            data = self.data2
        elif _index==3:
            data = self.data3
        else:
            return 0
        for i in range(0,4):
            intValue = intValue<<8
            intValue = intValue + data[3-i]
        return intValue
        
    def double2int32(self, dValue):
        result = int(dValue*100//1)
        return result
    
    #frame setter
    def setDataDF(self, _sub_type, _data_1, _data_2, _data_3):
        self.msg_type = 1;
        self.sub_type = _sub_type
        for i in range(0,4):
            self.data1[i] = (_data_1>>(8*i)) & 0x0FF
            self.data2[i] = (_data_2>>(8*i)) & 0x0FF
            self.data3[i] = (_data_3>>(8*i)) & 0x0FF
    
    def setOperationDF(self, _sub_type, _data_1, _data_2, _data_3):
        self.msg_type = 2;
        self.sub_type = _sub_type
        for i in range(0,4):
            self.data1[i] = (_data_1>>(8*i)) & 0x0FF
            self.data2[i] = (_data_2>>(8*i)) & 0x0FF
            self.data3[i] = (_data_3>>(8*i)) & 0x0FF
    
    def makeSendData(self):
        send_data_stream = bytearray()
        for i in range(0,2):
            send_data_stream.append(self.start[i])
        send_data_stream.append(self.msg_type)
        send_data_stream.append(self.sub_type)
        for i in range(0,4):
            send_data_stream.append(self.data1[i])
        for i in range(0,4):
            send_data_stream.append(self.data2[i])
        for i in range(0,4):
            send_data_stream.append(self.data3[i])
        send_data_stream.append(self.end)
        #ts test
        '''        
        for i in range(0, len(send_data_stream)):
            print ('%x' %send_data_stream[i] )
        '''
        return send_data_stream